metron-platform/metron-enrichment/src/main/java/org/apache/metron/enrichment/adapters/geo/GeoLiteDatabase.java

```diff
@@ -0,0 +1,184 @@
+    LOG.trace("[Metron] Determining if GeoIpDatabase update required");
+    String hdfsFile = GEO_HDFS_FILE_DEFAULT;
+    if (globalConfig != null) {
+      hdfsFile = (String) globalConfig.getOrDefault(GEO_HDFS_FILE, GEO_HDFS_FILE_DEFAULT);
+    }
+
+    // Always update if we don't have a DatabaseReader
+    if (reader == null || !hdfsLoc.equals(hdfsFile)) {
+      // Update
+      hdfsLoc = hdfsFile;
+      update(hdfsFile);
+    } else {
+      LOG.trace("[Metron] Update to GeoIpDatabase unnecessary");
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  public void update(String hdfsFile) {
+    // If nothing is set (or it's been unset, use the defaults)
+    if (hdfsFile == null || hdfsFile.isEmpty()) {
+      LOG.debug("[Metron] Using default for {}: {}", GEO_HDFS_FILE, GEO_HDFS_FILE_DEFAULT);
+      hdfsFile = GEO_HDFS_FILE_DEFAULT;
+    }
+
+    FileSystem fs;
+    try {
+      fs = FileSystem.get(new Configuration());
+    } catch (IOException e) {
+      LOG.error("[Metron] Unable to retrieve get HDFS FileSystem");
+      throw new IllegalStateException("[Metron] Unable to get HDFS FileSystem");
+    }
+
+    try (GZIPInputStream gis = new GZIPInputStream(fs.open(new Path(hdfsFile)))) {
+      writeLock.lock();
+      LOG.info("[Metron] Update to GeoIP data started with {}", hdfsFile);
+      // InputStream based DatabaseReaders are always in memory.
+      DatabaseReader newReader = new DatabaseReader.Builder(gis).withCache(new CHMCache()).build();
+      DatabaseReader oldReader = reader;
+      reader = newReader;
+      // If we've never set a reader, don't close the old one
+      if (oldReader != null) {
+        oldReader.close();
+      }
+      LOG.info("[Metron] Finished update to GeoIP data started with {}", hdfsFile);
+    } catch (IOException e) {
+      LOG.error("[Metron] Unable to open new database file {}", hdfsFile, e);
+      throw new IllegalStateException("[Metron] Unable to update MaxMind database");
+    } finally {
+      // Don't unlock if the try failed
+      if (lock.isWriteLocked()) {
+        writeLock.unlock();
+      }
+    }
+  }
+
+  // Optional.empty means that we don't have any geo location in database.
+  // Optional exists, but empty means local IP (valid, but no info will be in the DB)
+  @SuppressWarnings("unchecked")
+  public Optional<HashMap<String, String>> get(String ip) {
+    // Call get every single time, returns current version. Updates behind the scenes.
+    LOG.trace("[Metron] Called GeoIpDatabase.get({})", ip);
+    InetAddress addr = null;
+    try {
+      addr = InetAddress.getByName(ip);
+    } catch (UnknownHostException e) {
+      LOG.warn("[Metron] No result found for IP {}", ip, e);
+      return Optional.empty();
+    }
+    if (isLocalAddress(ip, addr)) {
+      return Optional.of(new HashMap());
+    }
+
+    try {
+      readLock.lock();
+      addr = InetAddress.getByName(ip);
+      CityResponse cityResponse = reader.city(addr);
```

Comment:
Would it make sense to constrain the scope of the read lock?  Once we have a `CityResponse` object we don't need the lock any longer.